### PR TITLE
Setup the product editor to support the product variation

### LINKED ISSUE #40573: Setup the product editor to support the product variation
**Prerequisites (mark completed items with an [x]):**
- [ ] I have checked that my issue type is not listed here https://github.com/woocommerce/woocommerce/issues/new/choose
- [ ] My issue is not a security issue, support request, bug report, enhancement or feature request (Please use the link above if it is).

**Issue Description:**

Before landing to a single variation page it's required to setup and configure the editor to work with product variations

AkVNGImLgSqCObTQ3idVn7-fi-7819_442831

**Epic**

https://github.com/woocommerce/woocommerce/issues/40304

**ACs**

- [ ] Create a new page called `ProductVariationPage` from the current `ProductPage`
- [ ] Register the route `/product/:productId/variation/:variationId` to render the `ProductVariationPage`
- [ ] The entity record provider needs to be configured to use `product` and  `product_variation` postTypes. [`show_in_rest`](https://github.com/woocommerce/woocommerce/blob/93b6f1b8ce5b27064ef9745cf4f49026365b114e/plugins/woocommerce/includes/class-wc-post-types.php#L367C36-L367C36) should be set for `product_variation`, so the entity record could request the variation from the server. The request will looks like `/wp/v2/product_variation` and it needs to be replaced by the [WC v3](https://woocommerce.github.io/woocommerce-rest-api-docs/#retrieve-a-product-variation). Consider to create a new middleware like `productApiFetchMiddleware` to do that.
- [ ] A new ProductVariationTemplate should be created and load into the editor.
- [ ] The template should has the `General` tab AkVNGImLgSqCObTQ3idVn7-fi-7819_444983
- [ ] The `General` tab should contains the `Variation details` section and the `Note` block AkVNGImLgSqCObTQ3idVn7-fi-7819_444984
- [ ] The `Note` block should be the same `Summary` block from the product editor
- [ ] Make sure the publish button work to save the variation

### PR DESCRIPTION
**Pull request by mdperez86**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

The base of this PR is adding support for using the product editor with multiple product types, as we previously had hardcoded the `product` post type in a lot of places. This would not work for loading a variation specific form.
In addition, this PR also adds a new page for loading a variation form and adds a new template for product variations.

**Note:**
This is the first of several PRs to make editing variations possible within the new editor, this PR updates most of the main items to get a product variation template loaded.

Closes https://github.com/woocommerce/woocommerce/issues/40573

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Create a WooCommerce store with this branch and enable the product editor using **WooCommerce > Settings > Advanced > Features**
2. Go to **Products > Add New**
3. Add a name and click save draft, now go to the variations tab and add several variation options.
4. Wait until the variations are generated.
5. Click on one of the 3 dot menus of a variation a take note of the **Variation Id** listed at the top.
6. Now update your URL removing the `&tab=variations` at the end and replacing it with `/variation/<variation_id>` (replacing the variation id with the one from the 3 dot menu )
7. This should load a new template with a General, Pricing, Inventory, and Shipping tab.
8. The General only contains a Variation Details and Images section. The variation details contains a note field.
9. Type something in the note field and click update
10. Updating should work correctly and the **Note** value should persist on refresh (Note: **Note** is the only field that will work at this moment ).

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

File: packages/js/product-editor/src/blocks/product-fields/regular-price/block.json [hunk 1]
```diff
@@ -25,5 +25,6 @@
 		"lock": false,
 		"__experimentalToolbar": false
 	},
+	"usesContext": [ "postType" ],
```
Comment (mattsherman) [inline packages/js/product-editor/src/blocks/product-fields/regular-price/block.json:28]: If every block is going to need this, maybe we should auto-inject it when the block is registered?

File: packages/js/product-editor/src/blocks/product-fields/summary/edit.tsx [hunk 2]
```diff
@@ -40,8 +41,8 @@ export function Edit( {
 	);
 	const [ summary, setSummary ] = useEntityProp< string >(
 		'postType',
-		'product',
-		'short_description'
+		context.postType || 'product',
+		attributes.name || 'short_description'
```
Comment (mattsherman) [inline packages/js/product-editor/src/blocks/product-fields/summary/edit.tsx]: I'm not sure I love the idea of specific blocks being reused like this with attributes controlling which property they are bound to. Can we make this block a bit more generic so that it can be reused by 3PDs as well?

I'd get rid of the default entity property binding and require the property to always be passed in.

File: packages/js/product-editor/src/blocks/product-fields/summary/block.json [hunk 3]
```diff
@@ -8,6 +8,9 @@
 	"keywords": [ "products", "summary", "excerpt" ],
 	"textdomain": "default",
 	"attributes": {
+		"name": {
+			"type": "string"
+		},
```
Comment (mattsherman) [inline packages/js/product-editor/src/blocks/product-fields/summary/block.json]: I'd suggest `property` as the attribute name. I think we should standardize of this for all property binding.

File: packages/js/product-editor/src/components/block-editor/block-editor.tsx [hunk 4]
```diff
@@ -36,18 +36,21 @@ import { ProductEditorContext } from '../../types';
 
 type BlockEditorProps = {
 	context: Partial< ProductEditorContext >;
-	product: Partial< Product >;
+	postType: 'product' | 'product_variation';
+	postId: number;
```
Comment (mattsherman) [inline packages/js/product-editor/src/components/block-editor/block-editor.tsx]: I know it is (currently) the post id, but I think it might be more clear to call it `productId`, since in the context of Woo, we are dealing with products, not posts. That they are also posts is largely an implementation detail that we have abstracted away in most of the Woo APIs.

File: packages/js/product-editor/src/components/editor/editor.tsx [hunk 5]
```diff
@@ -41,14 +42,21 @@ import { ValidationProvider } from '../../contexts/validation-context';
...
 	EditorSettings & EditorBlockListSettings
->;
+> & {
+	templates: Record< string, Template[] >;
+};
 
 type EditorProps = {
-	product: Product;
+	product: Product | ProductVariation;
+	postType?: 'product' | 'product_variation';
```
Comment (mattsherman) [inline packages/js/product-editor/src/components/editor/editor.tsx]: Can't we infer the `postType` from the `product` object?

What happens if there is a mismatch between `product` and `postType`?

File: packages/js/product-editor/src/components/editor/editor.tsx [hunk 6]
```diff
@@ -96,7 +106,10 @@ export function Editor( { product, settings }: EditorProps ) {
 					WooFooterItem to actually render in the WooFooterItem.Slot defined by
 					WooCommerce Admin. And, we need to put it outside of the SlotFillProvider
 					we create in this component. */ }
-					<Footer product={ product } />
+					<Footer
+						productId={ product.id }
+						productType={ product.type }
+					/>
```
Comment (mattsherman) [inline packages/js/product-editor/src/components/editor/editor.tsx:111]: Why change this instead of changing `Footer` to just look at the `.type` property of the `product` prop?

I think it's kinda nice when you can just pass along the entity instead of having to separately pass pieces of it. And, if you pass the entity, you can use additional parts of it in the future without having to change as much code.

File: packages/js/product-editor/src/utils/product-apifetch-middleware.ts [hunk 7]
```diff
@@ -4,6 +4,28 @@
...
+				'/product/([0-9]+)/variation/([0-9]+)'
+			);
+			const matched = ( query.path || '' ).match( variationMatcher );
+			if ( matched && matched.length === 3 ) {
+				return '/wc/v3/products/' + matched[ 1 ] + '/variations';
+			}
+			return '/wc/v3/products/0/variations';
+		},
+	},
+];
```
Comment (mattsherman) [inline packages/js/product-editor/src/utils/product-apifetch-middleware.ts]: Why is this needed? Oh wait... I think the comment in the code below this answers it.

File: packages/js/product-editor/src/utils/record-product-event.ts [hunk 8]
```diff
@@ -1,10 +1,13 @@
...
  * External dependencies
  */
-import { Product } from '@woocommerce/data';
+import { Product, ProductVariation } from '@woocommerce/data';
 import { recordEvent } from '@woocommerce/tracks';
 
-export function recordProductEvent( eventName: string, product: Product ) {
+export function recordProductEvent(
+	eventName: string,
+	product: Product | ProductVariation
```
Comment (mattsherman) [inline packages/js/product-editor/src/utils/record-product-event.ts]: It feels really unfortunate that we have to do this sort of thing all over the place.

File: packages/js/product-editor/src/utils/record-product-event.ts [hunk 9]
```diff
@@ -34,9 +37,9 @@ export function recordProductEvent( eventName: string, product: Product ) {
 
 	recordEvent( eventName, {
 		attributes: attributes.length ? 'yes' : 'no',
-		categories: categories.length ? 'yes' : 'no',
-		cross_sells: cross_sell_ids.length ? 'yes' : 'no',
-		description: description.length ? 'yes' : 'no',
+		categories: categories?.length ? 'yes' : 'no',
+		cross_sells: cross_sell_ids?.length ? 'yes' : 'no',
+		description: description?.length ? 'yes' : 'no',
```
Comment (mattsherman) [inline packages/js/product-editor/src/utils/record-product-event.ts]: Is it clear to mark these as `no` when the product type never has them? I guess it depends on how the tracks events are used for analysis.

And, do any of these get inherited from the parent variable product? If so, we really should record the inherited value.

File: plugins/woocommerce/includes/class-wc-post-types.php [hunk 10]
```diff
@@ -380,6 +380,7 @@ public static function register_post_types() {
 					'supports'        => false,
 					'capability_type' => 'product',
 					'rewrite'         => false,
+					'show_in_rest'    => true,
```
Comment (mattsherman) [inline plugins/woocommerce/includes/class-wc-post-types.php]: Does this have any side-effects that would be unexpected?

File: plugins/woocommerce/src/Admin/Features/ProductBlockEditor/Init.php [hunk 11]
```diff
@@ -77,11 +79,16 @@ public function enqueue_scripts() {
...
+		$template_registry 	  = wc_get_container()->get( BlockTemplateRegistry::class );
 
 		$editor_settings = array();
 		if ( ! empty( $post_type_object->template ) ) {
 			$editor_settings['template']                 = $post_type_object->template;
 			$editor_settings['templateLock']             = ! empty( $post_type_object->template_lock ) ? $post_type_object->template_lock : false;
+			$editor_settings['templates']     = array(
+				'product' 			=> $post_type_object->template,
+				'product_variation' => $template_registry->get_registered( 'product-variation' )->get_formatted_template()
+			);
```
Comment (mattsherman) [inline plugins/woocommerce/src/Admin/Features/ProductBlockEditor/Init.php]: As we mentioned, good enough for now until we can get the templates from the REST API.

File: plugins/woocommerce/src/Admin/Features/ProductBlockEditor/ProductTemplates/ProductVariationTemplate.php [hunk 12]
```diff
@@ -0,0 +1,574 @@
...
+ */
+
+namespace Automattic\WooCommerce\Admin\Features\ProductBlockEditor\ProductTemplates;
+
+use Automattic\WooCommerce\Admin\Features\Features;
+
+/**
+ * Simple Product Template.
+ */
+class ProductVariationTemplate extends AbstractProductFormTemplate implements ProductFormTemplateInterface {
```
Comment (mattsherman) [inline plugins/woocommerce/src/Admin/Features/ProductBlockEditor/ProductTemplates/ProductVariationTemplate.php:13]: I'd like to see us explore using conditional visibility when it is available to use the same template for both simple products and variations. Because, they seem very similar.

If not, we should carefully consider when we change ids for things like groups, sections, and fields that appear "the same" in both. Because, if they have the same ids in multiple templates, it will make extensibility easier (for the case where you want to do the same thing in all templates).

We should also consider using composition to build up both templates from common "building blocks" where they have similarity. This will at least allow us to not duplicate too much.

**Review by mattsherman — DISMISSED** [review]
Left a bunch of comments.

Overall, I don't love that we have to pass the product type all over the place. Is there a way we could get this via some sort of context?

It feels less than ideal that so many pieces of the codebase need to concern themselves with whether they are dealing with a product or a variation.

File: packages/js/product-editor/src/components/block-editor/block-editor.tsx (hunk 4 above)
Comment (louwie17) [reply packages/js/product-editor/src/components/block-editor/block-editor.tsx]: Same thing for `postType` and change that to `productType`?

File: packages/js/product-editor/src/components/editor/editor.tsx (hunk 5 above)
Comment (louwie17) [reply packages/js/product-editor/src/components/editor/editor.tsx]: Hmm infer would only work if we pass the specific product down into `Editor` already, which I suppose we are doing for Variation. From a Types standpoint that is, we could otherwise infer it and probably get it from the post type data.
Although this will not work when we support other product types, as we won't be creating a specific page for each one. Meaning that in order.

> What happens if there is a mismatch between product and postType?

This will mean that the right product won't be viewed, or it will retrieve another one from another endpoint.
Basically the `postType` and `postId` are the primary ids. So the redux store will looking under `postType > postId`. So if the post type is mismatched it will not find the right product.

File: plugins/woocommerce/src/Admin/Features/ProductBlockEditor/ProductTemplates/ProductVariationTemplate.php (hunk 12 above)
Comment (AnnaMag) [reply plugins/woocommerce/src/Admin/Features/ProductBlockEditor/ProductTemplates/ProductVariationTemplate.php:13]: > I'd like to see us explore using conditional visibility when it is available 

@mattsherman, I assume that [the relevant discussion](https://github.com/woocommerce/woocommerce/discussions/40399) has concluded FOR using conditional visibility to manage different product types. Is that right? Are you planning to wrap up/summarize the discussion or plan to wait for more input.

@louwie17 @mdperez86 
Should we move forward with implementing templates (since it is useful on its own) and migrating it to the alternative approach later, if decided? How much effort would it be to change it?
Given the ETA of conditional visibility, my concern is that it could become the blocker to shipping this work. 
Thoughts?

File: packages/js/product-editor/src/components/editor/editor.tsx (hunk 6 above)
Comment (louwie17) [reply packages/js/product-editor/src/components/editor/editor.tsx:111]: The main reason was for handling types, as I would also have to support a generic Post/Product type here.
Secondly, these two params will be enough for the component to retrieve its own product by using the data store. So I would almost prefer to have most components rely on these two parameters versus the entire object, and then just use a `useSelect` internally to determine what they do/don't need.

File: packages/js/product-editor/src/utils/product-apifetch-middleware.ts (hunk 7 above)
Comment (louwie17) [reply packages/js/product-editor/src/utils/product-apifetch-middleware.ts]: Yeah, this is kinda dirty!
I was also floating the idea of always passing an additional param like `useWCAPI: boolean` that it would use that over checking if we are on the product editing page.
And in theory we want to push some changes up stream that allows us to use the `useEntity` store with our own endpoints ( or at-least register some new one's client side, so we don't run the risk of overwriting other ones, as WC Blocks relies on the `wp/v2` endpoint for example.

File: packages/js/product-editor/src/utils/record-product-event.ts (hunk 9 above)
Comment (louwie17) [reply packages/js/product-editor/src/utils/record-product-event.ts]: Yeah this needs to be cleaned up!
For the sake of getting something working quickly, I resorted to the above. We did have to update this to be more flexible with other product types.

File: plugins/woocommerce/includes/class-wc-post-types.php (hunk 10 above)
Comment (louwie17) [reply plugins/woocommerce/includes/class-wc-post-types.php]: Good question, I was kinda curious about that to 🤷‍♂️ 
Last time we made changes here, we didn't find it out until after the code freeze and other people started complaining.

File: plugins/woocommerce/src/Admin/Features/ProductBlockEditor/Init.php (hunk 11 above)
Comment (louwie17) [reply plugins/woocommerce/src/Admin/Features/ProductBlockEditor/Init.php]: Yup, it does kinda suck this gets added to every WCAdmin page 😢

File: plugins/woocommerce/src/Admin/Features/ProductBlockEditor/ProductTemplates/ProductVariationTemplate.php (hunk 12 above)
Comment (louwie17) [reply plugins/woocommerce/src/Admin/Features/ProductBlockEditor/ProductTemplates/ProductVariationTemplate.php:13]: Yup, that would be cool.

Is it currently possible to create multiple `FormTemplates` for this, like a specific `FormTemplate` for a section? and use that across templates?

File: packages/js/product-editor/src/blocks/product-fields/summary/block.json (hunk 3 above)
Comment (louwie17) [reply packages/js/product-editor/src/blocks/product-fields/summary/block.json]: Yeah `property` may indeed be the better one to prevent with `name` being used for something else.
My only slight push back is that `name` is used for the native input fields and most form libraries, but I do see how this can be a bit of an exception.

File: packages/js/product-editor/src/blocks/product-fields/regular-price/block.json (hunk 1 above)
Comment (mdperez86) [reply packages/js/product-editor/src/blocks/product-fields/regular-price/block.json:28]: I think we are not going to need this if we get that value form the EntityContext (of course if it's exposed).

File: packages/js/product-editor/src/blocks/product-fields/summary/block.json (hunk 3 above)
Comment (mdperez86) [reply packages/js/product-editor/src/blocks/product-fields/summary/block.json]: Yep, I do agree with that, native input uses name to refers to an object prop and under the hood we should also use name in any of our form fields.

File: packages/js/product-editor/src/components/editor/editor.tsx (hunk 5 above)
Comment (mdperez86) [reply packages/js/product-editor/src/components/editor/editor.tsx]: The responsibility to render the right object is of the page and not of the editor, so we have to restrict which product should be handle by the page is being rendered. Also getting the object associated template should be load from the page depending on the object that is going to be edited and then passed to the editor as well as the object itself.

File: packages/js/product-editor/src/utils/record-product-event.ts (hunk 8 above)
Comment (mdperez86) [reply packages/js/product-editor/src/utils/record-product-event.ts]: True

File: plugins/woocommerce/src/Admin/Features/ProductBlockEditor/Init.php (hunk 11 above)
Comment (mdperez86) [reply plugins/woocommerce/src/Admin/Features/ProductBlockEditor/Init.php]: Yep, but we can do some logic here to recognize the type of product we are dealing with for the current request and only load the templates depending on that right?

File: plugins/woocommerce/src/Admin/Features/ProductBlockEditor/ProductTemplates/ProductVariationTemplate.php (hunk 12 above)
Comment (mdperez86) [reply plugins/woocommerce/src/Admin/Features/ProductBlockEditor/ProductTemplates/ProductVariationTemplate.php:13]: >Should we move forward with implementing templates (since it is useful on its own) and migrating it to the alternative approach later, if decided? How much effort would it be to change it?
Given the ETA of conditional visibility, my concern is that it could become the blocker to shipping this work.
Thoughts?

Yes, we are going to continue without the conditional block here. We don't have to much time to release this and decisions like that are going to be covered in further project and before the issues are being placed to InProgress. On the other hand we should identify first/better what is the ideal scenario to use conditional or template before starting to use it.

Comment (mdperez86) [reply plugins/woocommerce/src/Admin/Features/ProductBlockEditor/ProductTemplates/ProductVariationTemplate.php:13]: > Is it currently possible to create multiple FormTemplates for this, like a specific FormTemplate for a section? and use that across templates?

There are a lot of things we can do here to avoid duplications, I do agree with @mattsherman that a composition is a strong candidate here, but reusing or abstracting groups of blocks is going to require the design to be aligned across product types. This is of course and improvement topic we can discuss in more details in the future.

File: plugins/woocommerce/src/Admin/Features/ProductBlockEditor/Init.php (hunk 11 above)
Comment (mattsherman) [reply plugins/woocommerce/src/Admin/Features/ProductBlockEditor/Init.php]: I don't think so, because once in the product editor the routing is done client-side, so we won't hit this again.

File: packages/js/product-editor/src/blocks/product-fields/summary/block.json (hunk 3 above)
Comment (mattsherman) [reply packages/js/product-editor/src/blocks/product-fields/summary/block.json]: With native HTML input fields, the `name` property is not really the name of an object property it is bound to. It is the name of the field, and that happens to also get used to send the value of the form field to the server on a form submission.

`name` is vague and confusing, since we also have the block `name` (`woocommerce/product-summary`which is also poorly chosen, it should be `type` in my opinion). 

Especially with the support Nathan has been working on to support meta data as well, we might have property paths, like `meta_data.foo` bound to blocks in the future.

File: packages/js/product-editor/src/components/block-editor/block-editor.tsx (hunk 4 above)
Comment (mattsherman) [reply packages/js/product-editor/src/components/block-editor/block-editor.tsx]: I would.

File: plugins/woocommerce/src/Admin/Features/ProductBlockEditor/ProductTemplates/ProductVariationTemplate.php (hunk 12 above)
Comment (mattsherman) [reply plugins/woocommerce/src/Admin/Features/ProductBlockEditor/ProductTemplates/ProductVariationTemplate.php:13]: > I assume that https://github.com/woocommerce/woocommerce/discussions/40399 has concluded FOR using conditional visibility to manage different product types. Is that right? Are you planning to wrap up/summarize the discussion or plan to wait for more input.

We should include guidance around this in the UX/dev guidelines. I'll make a note to follow up.

> Is it currently possible to create multiple FormTemplates for this, like a specific FormTemplate for a section? and use that across templates?

Not currently (I have thought about 'template fragments'). Currently, the best that can be done is to use composition with PHP when building the template. Which isn't bad... I just think it could be simplified a bit via something like fragments.

> Yes, we are going to continue without the conditional block here. 

I think that is a reasonable decision in the short term. I don't think the effort to move to a conditional template would be that great if the two templates really are very similar (which is the scenario where a conditional template would make sense).

File: plugins/woocommerce/src/Admin/Features/ProductBlockEditor/Init.php (hunk 11 above)
Comment (mdperez86) [reply plugins/woocommerce/src/Admin/Features/ProductBlockEditor/Init.php]: > I don't think so, because once in the product editor the routing is done client-side, so we won't hit this again.

Right!

File: packages/js/product-editor/src/blocks/product-fields/summary/block.json (hunk 3 above)
Comment (mdperez86) [reply packages/js/product-editor/src/blocks/product-fields/summary/block.json]: > and that happens to also get used to send the value of the form field to the server on a form submission.

You said better than me.

File: plugins/woocommerce/includes/class-wc-post-types.php (hunk 10 above)
Comment (louwie17) [reply plugins/woocommerce/includes/class-wc-post-types.php]: I moved this to the Product Editor init file, so its only enabled when the product block editor is enabled.

File: packages/js/product-editor/src/blocks/product-fields/summary/block.json (hunk 3 above)
Comment (louwie17) [reply packages/js/product-editor/src/blocks/product-fields/summary/block.json]: Yup, makes sense. I did update the `summary` field to adhere to this.

**Review by octaedro — APPROVED** [review]
You did a great job here! LGTM 🚀
